syntax = "proto3";

package armonik.api.grpc.v1.results;

import "filters_common.proto";
import "google/protobuf/timestamp.proto";
import "result_status.proto";
import "results_fields.proto";

option csharp_namespace = "ArmoniK.Api.gRPC.V1.Results";

message FilterString {
  ResultField field = 1;
  string value = 2;
  FilterStringRange range = 3;
}

message FilterNumber {
  ResultField field = 1;
  int64 value = 2;
  FilterNumberRange range = 3;
}

message FilterDate {
  ResultField field = 1;
  google.protobuf.Timestamp value = 2;
  FilterDateRange range = 3;
}

message FilterBoolean {
  ResultField field = 1;
  bool value = 2;
}

message FilterArray {
  ResultField field = 1;
  string value = 2;
  FilterArrayRange range = 3;
}

message FilterStatus {
  ResultField field = 1;
  result_status.ResultStatus value = 2;
  FilterStatusRange range = 3;
}

message FilterField {
  oneof filter {
    FilterString string = 1;
    FilterNumber number = 2;
    FilterDate date = 3;
    FilterBoolean boolean = 4;
    FilterArray array = 5;
    FilterStatus status = 6;
  }
}

message FiltersAnd {
  repeated FilterField filters = 1;
}

message FiltersOr {
  repeated FiltersAnd filters = 1;
}

message Filters {
  FiltersOr filters = 1;
}

// TODO: I think that we could add a duration filter.
